Remove sstables::remove_by_toc_name()

xemul opened the pull request:
The helper in question complicates the logic of sstable_directory::process() by making garbage collection differently for sstables deleted "atomically" and deleted "one-by-one". Also, the code that deletes sstables one-by-one and uses remove_by_toc_name() renders excessive TOC file reading, because there's sstable object at hand and it had all_components() ready for use.

Surprisingly, there was no test for the deletion-log functionality. This PR adds one. The test passes before the g.c. and regular unlink fix, and (of course) continues passing after it.

bhalevy commented on sstables/sstable_directory.cc:
```diff
@@ -545,8 +545,8 @@ bool sstable_directory::compare_sstable_storage_prefix(const sstring& prefix_a,
 
-future<> sstable_directory::delete_with_pending_deletion_log(std::vector<shared_sstable> ssts) {
-    return seastar::async([ssts = std::move(ssts)] {
+future<std::pair<sstring, sstring>> sstable_directory::create_pending_deletion_log(const std::vector<shared_sstable>& ssts) {
```
please document the return value

bhalevy replied:
or maybe better, define a struct for it with named members.

bhalevy commented on sstables/sstable_directory.cc:
```diff
@@ -590,12 +590,12 @@ future<std::pair<sstring, sstring>> sstable_directory::create_pending_deletion_l
             close_out.close_now();
 
             auto dir_f = open_directory(pending_delete_dir).get0();
+            auto close_dir = deferred_close(dir_f);
```
nit: please spell check commit message.

Also, it may be more elegant to replace `dir_f.close().get()` with `close_dir.close_now()` so it becomes explicit and evident before the log message.

xemul replied:
Hm... OK

bhalevy commented on test/boost/sstable_directory_test.cc:
```diff
@@ -778,3 +778,53 @@ SEASTAR_THREAD_TEST_CASE(test_system_datadir_layout) {
     }, cfg).get();
 }
+
+SEASTAR_TEST_CASE(test_pending_log_garbage_collection) {
```
excellent, thanks!

bhalevy commented on sstables/sstables.hh:
```diff
@@ -1021,5 +1021,6 @@ future<> remove_table_directory_if_has_no_snapshots(fs::path table_dir);
 // Caller may pass sync_dir::no for batching multiple deletes in the same directory,
 // and make sure the directory is sync'ed on or after the last call.
 future<> remove_by_toc_name(sstring sstable_toc_name, storage::sync_dir sync = storage::sync_dir::yes);
+future<sstring> make_toc_temporary(sstring sstable_toc_name, storage::sync_dir sync = storage::sync_dir::yes);
```
Please document the return value semantics.

bhalevy commented on sstables/storage.cc:
```diff
@@ -409,7 +413,29 @@ future<> filesystem_storage::wipe(const sstable& sst, sync_dir sync) noexcept {
     try {
-        co_await remove_by_toc_name(name, sync);
+        auto new_toc_name = co_await make_toc_temporary(name, sync);
+        auto dir_name = parent_path(new_toc_name);
```
what if `new_toc_name == ""`?

xemul replied:
It would step on ENOENT from remove_file below and would anyway ignore this error. But you're right, this behavior was not exactly like that before the patch

xemul commented:
upd:
- rebased
- spell checking
- explicitly kick deferred-close before it fires on its own
- documented what values new helpers return
- fixed test debug-mode crash (@bhalevy , did you know that if `co_await get_units(sem)` goes sleeping on a semaphore from another shard it continues its execution on the semaphore's shard, not on the shard it was called on?)
- fixed wipe() to handle missing TOC explicitly

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5006/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5006/)
:white_check_mark: - [Sanity Tests](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6364/)
#### Build Details:
- Duration: 2 hr 39 min
- Builder: spider2.cloudius-systems.com

bhalevy commented:
> upd:
> 
> * rebased
> * spell checking
> * explicitly kick deferred-close before it fires on its own
> * documented what values new helpers return
> * fixed test debug-mode crash (@bhalevy , did you know that if `co_await get_units(sem)` goes sleeping on a semaphore from another shard it continues its execution on the semaphore's shard, not on the shard it was called on?)

@xemul Sounds familar... Since semaphore::wait enqueues a promise, this must be https://github.com/scylladb/seastar/issues/1563

> * fixed wipe() to handle missing TOC explicitly

bhalevy commented on sstables/sstable_directory.hh:
```diff
@@ -277,6 +277,10 @@ public:
     static future<> delete_with_pending_deletion_log(std::vector<shared_sstable> ssts);
+    // Creates the deletion log for atomic deletion of sstables (helper for the
+    // above function that's also used by tests)
+    // Returns a pair of "logilfe name" and "directory with sstables"
```
nit: logilfe -> logfile?

avikivity commented on sstables/sstable_directory.hh:
```diff
@@ -277,6 +277,10 @@ public:
+    // Creates the deletion log for atomic deletion of sstables (helper for the
+    // above function that's also used by tests)
+    // Returns a pair of "logilfe name" and "directory with sstables"
+    static future<std::pair<sstring, sstring>> create_pending_deletion_log(const std::vector<shared_sstable>& ssts);
```
Would have been better to return a struct.
